builder/builder.go

```diff
@@ -0,0 +1,606 @@
+		}
+		versionedBlockRequest = builderSpec.VersionedSubmitBlockRequest{
+			Version: spec.DataVersionCapella,
+			Capella: &blockSubmitReq,
+		}
+	case spec.DataVersionDeneb:
+		blockSubmitReq := builderApiDeneb.SubmitBlockRequest{
+			Signature:        signature,
+			Message:          blockBidMsg,
+			ExecutionPayload: payload.Deneb.ExecutionPayload,
+			BlobsBundle:      payload.Deneb.BlobsBundle,
+		}
+		versionedBlockRequest = builderSpec.VersionedSubmitBlockRequest{
+			Version: spec.DataVersionDeneb,
+			Deneb:   &blockSubmitReq,
+		}
+	}
+	return &versionedBlockRequest, err
+}
+
+func (b *Builder) handlePayloadAttributes(attrs *BuilderPayloadAttributes) error {
+	log.Info("handling payload attribute", "slot", attrs.Slot, "hash", attrs.HeadHash)
+	if attrs == nil {
+		return nil
+	}
+
+	parentBlock := b.eth.GetBlockByHash(attrs.HeadHash)
+	if parentBlock == nil {
+		return fmt.Errorf("parent block hash not found in block tree given head block hash %s", attrs.HeadHash)
+	}
+
+	proposerPubkey := phase0.BLSPubKey{}
+
+	if !b.eth.Synced() {
+		return errors.New("backend not Synced")
+	}
+
+	b.slotMu.Lock()
+	defer b.slotMu.Unlock()
+
+	if attrs.Equal(&b.slotAttrs) {
+		log.Debug("ignoring known payload attribute", "slot", attrs.Slot, "hash", attrs.HeadHash)
+		return nil
+	}
+
+	if b.slotCtxCancel != nil {
+		b.slotCtxCancel()
+	}
+
+	slotCtx, slotCtxCancel := context.WithTimeout(context.Background(), b.builderBlockTime)
+	b.slotAttrs = *attrs
+	b.slotCtx = slotCtx
+	b.slotCtxCancel = slotCtxCancel
+
+	go b.runBuildingJob(b.slotCtx, proposerPubkey, attrs)
+	return nil
+}
+
+func (b *Builder) runBuildingJob(slotCtx context.Context, proposerPubkey phase0.BLSPubKey, attrs *BuilderPayloadAttributes) {
+	ctx, cancel := context.WithTimeout(slotCtx, b.builderBlockTime)
+	defer cancel()
+
+	// Submission queue for the given payload attributes
+	// multiple jobs can run for different attributes fot the given slot
+	// 1. When new block is ready we check if its profit is higher than profit of last best block
+	//    if it is we set queueBest* to values of the new block and notify queueSignal channel.
+	var (
+		queueMu                sync.Mutex
+		queueLastSubmittedHash common.Hash
+		queueBestBlockValue    *big.Int = big.NewInt(0)
+	)
+
+	log.Info("runBuildingJob", "slot", attrs.Slot, "parent", attrs.HeadHash, "payloadTimestamp", uint64(attrs.Timestamp), "txs", attrs.Transactions)
+
+	// retry build block every builderBlockRetryInterval
+	runRetryLoop(ctx, b.builderRetryInterval, func() {
+		log.Info("retrying BuildBlock",
+			"slot", attrs.Slot,
+			"parent", attrs.HeadHash,
+			"retryInterval", b.builderRetryInterval.String())
+		payload, err := b.eth.BuildBlock(attrs)
+		if err != nil {
+			log.Warn("Failed to build block", "err", err)
+			return
+		}
+
+		sealedAt := time.Now()
+		queueMu.Lock()
+		defer queueMu.Unlock()
+		if payload.ExecutionPayload.BlockHash != queueLastSubmittedHash && payload.BlockValue.Cmp(queueBestBlockValue) > 0 {
```

Comment:
if the block is empty, the `payload.BlockValue` is 0. 
payload.BlockValue.Cmp(queueBestBlockValue) > 0 should be payload.BlockValue.Cmp(queueBestBlockValue) >= 0 to account for empty blocks

Reply:
I fixed it to `payload.BlockValue.Cmp(queueBestBlockValue) >= 0`